File: 04-computation/continuing14_20260908_lrc_zero_classification.cpp
#include <algorithm>
#include <array>
#include <fstream>
#include <iostream>
#include <map>
#include <numeric>
#include <set>
#include <string>
#include <unordered_set>
#include <vector>
using namespace std; using I=long long; using Word=array<int,7>;
I gates=0,raw_words=0,evals=0;
void need(bool ok,string message){++gates;if(!ok){cerr<<message<<'\n';exit(1);}}
I mod(I a,I b){I r=a%b;return r<0?r+b:r;}
I floorq(I a,I b){return (a-mod(a,b))/b;}
struct Geo{int p,q;I L;vector<pair<I,I>> arcs;};
bool allowed(int s){for(int p=2;p*p<=s;p++){int e=0;while(s%p==0){s/=p;e++;}if(e&&(p%3!=2||e>2))return false;}return s==1||s%3==2;}
Geo geometry(int p,int q){
 Geo G{p,q,14LL*p*q,{}};vector<pair<I,I>> A,B,C;
 for(int k=0;k<=p;k++)A.push_back({max(0LL,(14LL*k-1)*q),min(G.L,(14LL*k+1)*q)});
 for(int k=0;k<=q;k++)B.push_back({max(0LL,(14LL*k-1)*p),min(G.L,(14LL*k+1)*p)});
 size_t i=0,j=0;while(i<A.size()&&j<B.size()){
  I a=max(A[i].first,B[j].first),b=min(A[i].second,B[j].second);if(a<b)C.push_back({a,b});
  if(A[i].second<B[j].second)i++;else if(A[i].second>B[j].second)j++;else{i++;j++;}
 }
 G.arcs.push_back({C.back().first-G.L,C.front().second});G.arcs.insert(G.arcs.end(),C.begin()+1,C.end()-1);return G;
}
pair<I,I> capacity(int n,const Geo&G){
 I base=0,cur=0;map<I,array<int,3>> events;
 for(auto[a,b]:G.arcs){I h=n*(b-a)/G.L,r=n*(b-a)%G.L,A=mod(n*a,G.L),B=mod(n*b,G.L);base+=h;
  if(!r)events[A][2]++;else{events[A][0]++;events[B][1]++;cur+=A>B;}}
 I walls=1000000000,chambers=1000000000;
 for(auto[e,v]:events){walls=min(walls,cur-v[1]-v[2]);cur+=v[0]-v[1];chambers=min(chambers,cur);}
 return {base+min(walls,chambers),base+chambers};
}
I direct(int n,const Geo&G,I phase2){I total=0;for(auto[a,b]:G.arcs)total+=-floorq(-2LL*n*b+phase2,2*G.L)-floorq(2LL*n*a-phase2,2*G.L)-1;return total;}
bool danger(I x,I den,int p){I z=mod(p*x,den);return 14*min(z,den-z)<den;}
I literal(int n,const Geo&G,I phase2){I total=0;for(int j=0;j<n;j++){I x=2*G.L*j+phase2,den=2*G.L*n;total+=danger(x,den,G.p)&&danger(x,den,G.q);}return total;}
void controls(){
 for(auto[p,q]:vector<pair<int,int>>{{1,1},{1,13},{1,112},{1,355},{11,263},{23,323},{33,320},{48,307}}){auto G=geometry(p,q);
  for(int n:{7,14,29,651,1260,1890,1939,1953,2485}){set<I>walls;for(auto[a,b]:G.arcs){walls.insert(mod(n*a,G.L));walls.insert(mod(n*b,G.L));}
   I closed=1000000000,open=1000000000,phase=0;for(I w:walls){I a=direct(n,G,2*w),b=direct(n,G,2*w+1);if(min(a,b)<closed){closed=min(a,b);phase=a<=b?2*w:2*w+1;}open=min(open,b);}
   need(capacity(n,G)==make_pair(closed,open),"all-wall and all-chamber literal integer interval control");need(literal(n,G,phase)==closed,"literal native-grid control");
  }
 }
 need(capacity(1890,geometry(48,307)).second==0,"old7560 absolute-depth hostile survives chamber refinement");
 need(capacity(1953,geometry(33,320)).second==0,"new7812 singleton slot is individually realizable");
 need(gcd(7812,1280)==4&&gcd(7812,132)==12,"new7812 singleton arm has exact actual margins");
}
int gd[91][91];array<unordered_set<unsigned long long>,7> profiles;vector<int> masks;
bool valid(const Word&w){
 int g=0;for(int d:w)g=gd[g][d];if(g!=1)return false;
 int cache[128]={};auto get=[&](auto&&self,int m)->int{if(cache[m])return cache[m];int bit=__builtin_ctz((unsigned)m),r=m&(m-1);return cache[m]=r?gd[w[bit]][self(self,r)]:w[bit];};
 for(int m:masks){int c=get(get,m),k=7-__builtin_popcount((unsigned)m),j=0;array<int,6>A{};
  for(int i=0;i<7;i++)if(!(m&(1<<i)))A[j++]=gd[c][w[i]];sort(A.begin(),A.begin()+k);
  unsigned long long key=c;for(int i=0;i<k;i++)key=(key<<7)|A[i];if(!profiles[k].count(key))return false;
 }return true;
}
void enumerate(const vector<int>&D,Word&w,int pos,int first,vector<Word>&out){
 if(pos==7){raw_words++;if(valid(w))out.push_back(w);return;}
 for(int j=first;j<(int)D.size();j++){w[pos]=D[j];enumerate(D,w,pos+1,j,out);}
}
using Graph=array<array<unsigned char,91>,91>;
bool connected(const Word&w,const Graph&G,int bit){int reached=1;
 while(true){int old=reached;for(int i=0;i<7;i++)if(reached&(1<<i))for(int j=0;j<7;j++)if(i!=j&&(G[w[i]][w[j]]&bit))reached|=1<<j;if(old==reached)return reached==127;}
}
void writeword(ostream&o,const Word&w){for(int i=0;i<7;i++)o<<' '<<w[i];}
int main(int argc,char**argv){
 need(argc==3,"input/output paths required");ifstream in(argv[1]);string dest=argv[2];
 for(int a=0;a<=90;a++)for(int b=0;b<=90;b++)gd[a][b]=gcd(a,b);
 int np;in>>np;for(int i=0;i<np;i++){int k,c;in>>k>>c;unsigned long long key=c;for(int j=0;j<k;j++){int x;in>>x;key=(key<<7)|x;}profiles[k].insert(key);}
 for(int k=6;k>=1;k--)for(int m=1;m<127;m++)if(__builtin_popcount((unsigned)m)==k)masks.push_back(m);need(masks.size()==126,"complete126 masks");
 int nd;in>>nd;vector<vector<int>> domains(nd);for(auto&D:domains){int d;in>>d;D.resize(d);for(int&x:D)in>>x;}
 int nt;in>>nt;vector<pair<int,int>> clocks(nt);set<int> quotients;
 for(auto&[t,d]:clocks){in>>t>>d;need(t%7==0,"zero budget clock");for(int e:domains[d]){need(t%(7*e)==0,"whole alphabet zero budget");quotients.insert(t/e);}}
 need(bool(in),"input complete");controls();vector<Geo> AT;for(int s=3;s<=356;s++)if(allowed(s))for(int p=1;p*2<s;p++)if(gcd(p,s-p)==1)AT.push_back(geometry(p,s-p));need(AT.size()==5855,"full native atlas");
 map<int,Graph> zero;map<int,array<array<int,91>,91>> slots;ofstream zf(dest+"/quotient_zero_pairs.txt");
 ofstream bf(dest+"/critical_banks.txt");set<int>critical={1939,1953};for(int a:{4,8,9,16,18,24})for(int b:{4,8,9,16,18,24})critical.insert(7056/gcd(a,b));
 for(int n:quotients){Graph G{};I local=0;
  array<array<int,91>,91> closed_counts{},open_counts{};
  for(auto&A:AT){int a=gcd(n,A.p),b=gcd(n,A.q);if(a>90||b>90||a%7==0||b%7==0)continue;if(a>b)swap(a,b);
   I lower=0;for(auto[x,y]:A.arcs)lower+=floorq((I)n*(y-x)-1,A.L);if(lower>0)continue;
   auto[c,o]=capacity(n,A);evals++;local++;need(c>=0&&o>=c,"wall/chamber capacity order");
   int bits=(c==0?1:0)|(o==0?2:0);if(bits){G[a][b]|=bits;G[b][a]|=bits;}
   if(bits&&critical.count(n))bf<<n<<' '<<A.p<<' '<<A.q<<' '<<gcd(n,A.p)<<' '<<gcd(n,A.q)<<' '<<bits<<'\n';
   if(c==0)closed_counts[a][b]+=a==b?2:1;
   if(o==0)open_counts[a][b]+=a==b?2:1;
  }
  for(int a=1;a<=90;a++)for(int b=a;b<=90;b++)if(G[a][b]){zf<<n<<' '<<a<<' '<<b<<' '<<(int)G[a][b]<<' '<<closed_counts[a][b]<<' '<<open_counts[a][b]<<'\n';open_counts[b][a]=open_counts[a][b];}zero.emplace(n,G);slots.emplace(n,open_counts);
 }
 cerr<<"QUOTIENTS "<<quotients.size()<<" EVALUATIONS "<<evals<<'\n';
 vector<vector<Word>> words(nd);ofstream df(dest+"/domain_counts.txt");
 for(int d=0;d<nd;d++){I before=raw_words;Word w{};enumerate(domains[d],w,0,0,words[d]);df<<d<<' '<<words[d].size()<<' '<<raw_words-before<<'\n';
  ofstream wf(dest+"/words_"+to_string(d)+".json");wf<<'[';bool first=true;for(auto&v:words[d]){if(!first)wf<<',';first=false;wf<<'[';for(int j=0;j<7;j++){if(j)wf<<',';wf<<v[j];}wf<<']';}wf<<']';
 }
 ofstream out(dest+"/survey_clocks.txt");int closed=0,openclosed=0,slotclosed=0;I totalwords=0;
 ofstream residual(dest+"/residual_7056.txt");
 for(auto[t,d]:clocks){Graph G{};for(int a:domains[d])for(int b:domains[d]){int e=gd[a][b];G[a][b]=zero.at(t/e)[a/e][b/e];}
  I cw=0,ow=0,sw=0;Word fc{},fo{},fs{};
  for(auto&w:words[d]){totalwords++;if(connected(w,G,1)){if(cw==0)fc=w;cw++;}if(connected(w,G,2)){if(ow==0)fo=w;ow++;
    array<int,91> mult{};for(int a:w)mult[a]++;bool possible=true;
    for(int a:domains[d])if(mult[a]&&!(G[a][a]&2)){int bound=0;for(int b:domains[d])if(a!=b&&mult[b]){int e=gd[a][b];bound+=mult[b]*slots.at(t/e)[a/e][b/e];}if(mult[a]>bound)possible=false;}
    if(possible){if(sw==0)fs=w;sw++;if(t==7056){for(int a:w)residual<<a<<' ';residual<<'\n';}}
  }}
  need(ow<=cw,"chamber zero graph contained in closed-wall zero graph");closed+=cw==0;openclosed+=ow==0;
  need(sw<=ow,"distinct-speed slot bound only removes chamber residuals");slotclosed+=sw==0;
  out<<t<<' '<<d<<' '<<words[d].size()<<' '<<cw<<' '<<ow<<' '<<sw;writeword(out,fc);writeword(out,fo);writeword(out,fs);out<<'\n';
 }
 cout<<"CLOCKS "<<nt<<" QUOTIENTS "<<quotients.size()<<" DOMAINS "<<nd<<"\nRAW_WORDS "<<raw_words<<" WORD_CLOCKS "<<totalwords<<"\nCLOSED_WALL_CLOSURES "<<closed<<" CHAMBER_CLOSURES "<<openclosed<<" SLOT_CLOSURES "<<slotclosed<<"\nCAPACITY_EVALUATIONS "<<evals<<" GATES "<<gates<<'\n';
}

File: 04-computation/continuing14_20260908_lrc_zero_classification.py
"""Complete zero-budget graph classification and two retained-coordinate refinements.

The native companion exhausts all 126-profile words. This wrapper pins the
universe, preserves all three deletion stages, and independently verifies the
18 chamber-only and 31 slot-only complete residuals and the final stop bank.
"""
from pathlib import Path
from math import gcd,comb
from collections import Counter
from itertools import combinations
from hashlib import sha256
import json,subprocess,sys,tempfile

sys.stdout.reconfigure(encoding='utf-8',newline='\n')
HERE=Path(__file__).resolve()
ROOT=HERE.parent.parent if HERE.parent.name=='04-computation' else Path.cwd()
DEST=ROOT/'05-knowledge/results' if HERE.parent.name=='04-computation' else HERE.parent
PROFILE_SHA='935f3f687b6d7c89cc099e536f536238fd753bcc4c1747906d213cef387ca93f'
BASELINE_SHA='89a3e7545cc77467c5f85fbe0bdaab1f71071cff65184bbc1212eb86c9f07177'
gates=0
def need(ok,why):
    global gates
    gates+=1
    if not ok:raise ArithmeticError(why)
def canonical(v):return json.dumps(v,sort_keys=True,separators=(',',':')).encode()
def semantic(v):return sha256(canonical(v)).hexdigest()
def matrix(t,word,table):
    out={}
    for a in set(word):
        for b in set(word):
            e=gcd(a,b);p=tuple(sorted((a//e,b//e)))
            out[a,b]=table.get((t//e,*p),(0,0,0))
    return out
def components(word,M,bit):
    todo=set(range(7));out=[]
    while todo:
        reached={min(todo)};old=set()
        while reached!=old:
            old=set(reached)
            reached|={j for i in old for j in range(7) if j!=i and M[word[i],word[j]][0]&bit}
        todo-=reached;out.append(sorted(reached))
    return sorted(out)
def slot_witness(word,M):
    mult=Counter(word)
    for a in sorted(mult):
        if M[a,a][0]&2:continue
        terms=[[b,mult[b],M[a,b][2]] for b in sorted(mult) if b!=a and M[a,b][2]]
        bound=sum(n*r for b,n,r in terms)
        if mult[a]>bound:return dict(margin=a,multiplicity=mult[a],slot_bound=bound,terms=terms)
    return None
def geometry(p,q):
    L=14*p*q;A=[(max(0,(14*k-1)*q),min(L,(14*k+1)*q)) for k in range(p+1)]
    B=[(max(0,(14*k-1)*p),min(L,(14*k+1)*p)) for k in range(q+1)]
    out=[];i=j=0
    while i<len(A) and j<len(B):
        a,b=max(A[i][0],B[j][0]),min(A[i][1],B[j][1])
        if a<b:out.append((a,b))
        if A[i][1]<B[j][1]:i+=1
        elif A[i][1]>B[j][1]:j+=1
        else:i+=1;j+=1
    return L,[(out[-1][0]-L,out[0][1])]+out[1:-1]
def literal_control(n,p,q,chamber):
    L,I=geometry(p,q);walls=sorted({n*x%L for ab in I for x in ab})
    phases=[2*w+1 for w in walls] if chamber else [2*w for w in walls]
    def count(r):return sum(-((-(2*n*b-r))//(2*L))-(2*n*a-r)//(2*L)-1 for a,b in I)
    r=min(phases,key=count)
    def danger(j,v):
        den=2*L*n;a=(v*(2*L*j+r))%den
        return 14*min(a,den-a)<den
    A={j for j in range(n) if danger(j,p)};B={j for j in range(n) if danger(j,q)}
    need(len(A&B)==count(r),'literal native-grid intersection agrees with interval arithmetic')
    return dict(quotient_clock=n,p=p,q=q,phase2=r,phase_denominator=2*L,intersection=len(A&B),danger_sizes=[len(A),len(B)],marginal_bound=n//7)

def main():
    raw=(ROOT/'04-computation/overnight12_20260906_lrc_decoder_descent_inherited_profiles.json').read_bytes()
    need(sha256(raw).hexdigest()==PROFILE_SHA,'full profile supplier pin')
    P=json.loads(raw);V=P['levels']['6']['gcds']
    baseline=json.loads((ROOT/'05-knowledge/results/continuing13_20260908_lrc_zero_cuts_certificate.json').read_bytes())['new_scales']
    need(semantic(baseline)==BASELINE_SHA and len(baseline)==7618 and max(baseline)==11935,'complete inherited necessary array')
    clocks=[t for t in baseline if t%7==0]
    need(len(clocks)==1085,'declared complete remaining zero-budget universe')
    need(len(V)==42 and all(d%7 for d in V),'whole inherited alphabet has no seven factor')
    domains=sorted({tuple(d for d in V if t%d==0) for t in clocks})
    need(len(domains)==203,'all full divisor alphabets')
    index={D:i for i,D in enumerate(domains)}
    rows=[(k,c,w) for k,L in P['levels'].items() for c,w in L['profiles']]
    lines=[str(len(rows))]+[' '.join(map(str,[k,c,*w])) for k,c,w in rows]
    lines += [str(len(domains))]+[' '.join(map(str,[len(D),*D])) for D in domains]
    lines += [str(len(clocks))]+[str(t)+' '+str(index[tuple(d for d in V if t%d==0)]) for t in clocks]
    cpp=HERE.with_suffix('.cpp')
    with tempfile.TemporaryDirectory(prefix='continuing14_lrc_') as directory:
        tmp=Path(directory).resolve()
        need(tmp.parent==Path(tempfile.gettempdir()).resolve(),'private temporary output is under the resolved system temporary directory')
        inp=tmp/'input.txt';inp.write_bytes(('\n'.join(lines)+'\n').encode())
        exe=tmp/('native.exe' if sys.platform=='win32' else 'native')
        build=subprocess.run(['g++','-O2','-std=c++17',str(cpp),'-o',str(exe)],capture_output=True)
        need(build.returncode==0,'native compiler: '+build.stderr.decode(errors='replace'))
        run=subprocess.run([str(exe),str(inp),str(tmp)],capture_output=True)
        need(run.returncode==0,'native complete consumer: '+run.stderr.decode(errors='replace'))
        native=run.stdout.decode().replace('\r\n','\n')
        need('RAW_WORDS 9978787 WORD_CLOCKS 630818' in native,'whole unpruned universe and valid evaluations')
        need('CLOSED_WALL_CLOSURES 713 CHAMBER_CLOSURES 721 SLOT_CLOSURES 729' in native,'all three separately retained closure counts')
        domain_rows=[list(map(int,l.split())) for l in (tmp/'domain_counts.txt').read_text().splitlines()]
        wordbanks=[];domain_cert=[]
        for d,count,raw_count in domain_rows:
            data=(tmp/f'words_{d}.json').read_bytes();words=json.loads(data)
            need(len(words)==count and raw_count==comb(len(domains[d])+6,7),'domain whole multiset universe and accepted bank size')
            wordbanks.append(words)
            domain_cert.append(dict(domain=list(domains[d]),word_count=count,unpruned_multisets=raw_count,words_sha256=sha256(data).hexdigest()))
        need(sum(r['word_count'] for r in domain_cert)==415506,'distinct complete accepted profile words')
        zero_rows=[list(map(int,l.split())) for l in (tmp/'quotient_zero_pairs.txt').read_text().splitlines()]
        table={(n,a,b):(bits,closed,opened) for n,a,b,bits,closed,opened in zero_rows}
        need(len(table)==728,'complete nonempty quotient pair bank')
        for n,a,b,bits,c,o in zero_rows:
            need(0<=o<=c and bool(bits&1)==bool(c) and bool(bits&2)==bool(o),'oriented zero-bank counts and graph flags agree')
            need(gcd(a,b)==1 and n%7==0,'native reduced margin typing')
        records=[list(map(int,l.split())) for l in (tmp/'survey_clocks.txt').read_text().splitlines()]
        need([r[0] for r in records]==clocks,'every predeclared clock appears exactly once in order')
        plain=[r[0] for r in records if r[3]==0]
        chamber=[r[0] for r in records if r[3] and r[4]==0]
        slot=[r[0] for r in records if r[4] and r[5]==0]
        need(chamber==[1855,1911,1939,1981,2051,2121,3822,4242],'exact extra marginal-attainment clock bank')
        need(slot==[3738,3906,4074,5712,5964,6216,7308,7812],'exact extra distinct-speed slot bank')
        refinements=0
        for T,D,n,c,o,s,*_ in records:
            for t,d,nn,cc,oo,ss,*_ in records:
                if t>=T or T%t or d!=D:continue
                refinements+=1
                need(c<=cc and o<=oo and s<=ss,'unchanged-alphabet divisible refinement cannot add a residual word')
                small=matrix(t,domains[d],table);large=matrix(T,domains[d],table)
                for a in domains[d]:
                    for b in domains[d]:
                        need(large[a,b][1]<=small[a,b][1] and large[a,b][2]<=small[a,b][2],'complete oriented zero-slot counts decrease under lawful divisible refinement')
        selected=set(chamber+slot+[7056]);details=[]
        for t,d,n,c,o,s,*owners in records:
            need(n==len(wordbanks[d]) and 0<=s<=o<=c<=n,'complete clock-domain linkage and residual nesting')
            if t not in selected:continue
            residual_c=[];residual_o=[];residual_s=[];proof=[]
            for w in wordbanks[d]:
                M=matrix(t,w,table);cc=components(w,M,1);oo=components(w,M,2)
                if len(cc)==1:residual_c.append(w)
                if len(oo)!=1:
                    if len(cc)==1:proof.append(dict(word=w,chamber_components=oo))
                    continue
                residual_o.append(w);bad=slot_witness(w,M)
                if bad:proof.append(dict(word=w,slot_failure=bad))
                else:residual_s.append(w)
            need([len(residual_c),len(residual_o),len(residual_s)]==[c,o,s],'independent Python reconstruction of every refined-clock complete residual')
            details.append(dict(clock=t,closed_residuals=residual_c,chamber_residuals=residual_o,slot_residuals=residual_s,proofs=proof))
        stop=next(r for r in details if r['clock']==7056)
        need(stop['slot_residuals']==[[4,8,8,9,9,18,24],[4,8,9,9,16,18,24],[4,9,9,16,16,18,24]],'exact complete maximal remaining E0-clock residual')
        critical=[list(map(int,l.split())) for l in (tmp/'critical_banks.txt').read_text().splitlines()]
        need([r for r in critical if r[0]==1939]==[[1939,13,321,1,1,1]],'complete wall-only hostile bank')
        need([r for r in critical if r[0]==1953]==[[1953,33,320,3,1,3]],'complete singleton slot bank')
        wall=literal_control(1939,13,321,False)
        good=literal_control(1953,33,320,True)
        need(wall['intersection']==0 and min(wall['danger_sizes'])<277,'wall zero loses marginal attainment')
        need(good['intersection']==0 and good['danger_sizes']==[279,279],'singleton native arm has a saturated zero chamber')
        need(table[(1953,1,3)][2]==1 and gcd(7812,1280)==4 and gcd(7812,132)==12,'one slot is physically attainable but cannot serve three distinct speeds')
        old=literal_control(1890,48,307,True)
        need(old['intersection']==0 and old['danger_sizes']==[270,270],'old7560 depth hostile retains saturated individual zero arms')
        old_counts=[];removed=[]
        for name,stage in [('ordinary_zero_cuts',plain),('marginal_attainment',chamber),('distinct_speed_slots',slot)]:
            removed+=stage;remaining=[t for t in baseline if t not in set(removed)]
            old_counts.append(dict(mechanism=name,removed_clocks=stage,new_scales=remaining,new_scales_sha256=semantic(remaining)))
        new=old_counts[-1]['new_scales'];zero_left=[t for t in new if t%7==0]
        need(len(new)==6889 and max(new)==11934,'exact final general necessary array')
        need(len(zero_left)==356 and max(zero_left)==7056,'exact final zero-budget slice')
        cert=dict(status='PROVED analytic zero-budget consumers; FINITE-EXACT complete remaining E0 classification',
          profile_sha256=PROFILE_SHA,baseline_semantic_sha256=BASELINE_SHA,declared_clocks=clocks,
          native_source_sha256=sha256(cpp.read_bytes()).hexdigest(),native_transcript=native,
          domains=domain_cert,quotient_zero_pair_columns=['n','a','b','graph_bits','oriented_closed_count','oriented_chamber_count'],
          quotient_zero_pairs=zero_rows,clock_columns=['t','domain','word_count','closed_residual_count','chamber_residual_count','slot_residual_count','first_closed_word[7]','first_chamber_word[7]','first_slot_word[7]'],
          clocks=records,stages=old_counts,refined_complete_residuals=details,critical_native_banks=critical,
          controls=[wall,good,old],new_scales=new,new_scales_sha256=semantic(new),remaining_E0=zero_left,
          lawful_divisible_refinement_pairs=refinements,gates=gates,scope='Primitive thirteen distinct speeds; selected-six gcd clock; actual strict graph on complementary seven labels connected; all126 inherited profiles. Weak clearance only. No assertion that a remaining graph or slot assignment has an actual simultaneous ratio/phase realization.')
    data=canonical(cert)+b'\n';(DEST/(HERE.stem+'_certificate.json')).write_bytes(data)
    print('Complete E=0 clock classification: PASS')
    print(native.rstrip())
    print('SEPARATE NEW CLOSURES: ordinary zero cuts713; marginal attainment8; distinct-speed slots8')
    print('NECESSARY CLOCKS6889; maximum11934; remaining E0 clocks356; E0 maximum7056')
    print('STOP: full three-word7056 graph/slot residual retained; no joint ratio or phase realization inferred')
    print('NEW_SCALES_SHA256',semantic(new))
    print('CERTIFICATE_SHA256',sha256(data).hexdigest())
    print('Python always-active gates:',gates)
if __name__=='__main__':main()
